cpp/src/arrow/flight/flight-sql/client_test.cc

```diff
@@ -0,0 +1,569 @@
+                                      const FlightDescriptor& descriptor,
+                                      std::unique_ptr<FlightInfo>* info) {
+  return client->GetFlightInfo(options, descriptor, info);
+}
+
+Status FlightClientImpl_DoPut(FlightClientImpl* client, const FlightCallOptions& options,
+                              const FlightDescriptor& descriptor,
+                              const std::shared_ptr<Schema>& schema,
+                              std::unique_ptr<FlightStreamWriter>* stream,
+                              std::unique_ptr<FlightMetadataReader>* reader) {
+  return client->DoPut(options, descriptor, schema, stream, reader);
+}
+
+Status FlightClientImpl_DoGet(FlightClientImpl* client, const FlightCallOptions& options,
+                              const Ticket& ticket,
+                              std::unique_ptr<FlightStreamReader>* stream) {
+  return client->DoGet(options, ticket, stream);
+}
+
+Status FlightClientImpl_DoAction(FlightClientImpl* client,
+                                 const FlightCallOptions& options, const Action& action,
+                                 std::unique_ptr<ResultStream>* results) {
+  return client->DoAction(options, action, results);
+}
+
+}  // namespace internal
+
+class FlightMetadataReaderMock : public FlightMetadataReader {
+ public:
+  std::shared_ptr<Buffer>* buffer;
+
+  explicit FlightMetadataReaderMock(std::shared_ptr<Buffer>* buffer) {
+    this->buffer = buffer;
+  }
+
+  Status ReadMetadata(std::shared_ptr<Buffer>* out) override {
+    *out = *buffer;
+    return Status::OK();
+  }
+};
+
+class FlightStreamWriterMock : public FlightStreamWriter {
+ public:
+  FlightStreamWriterMock() = default;
+
+  Status DoneWriting() override { return Status::OK(); }
+
+  Status WriteMetadata(std::shared_ptr<Buffer> app_metadata) override {
+    return Status::OK();
+  }
+
+  Status Begin(const std::shared_ptr<Schema>& schema,
+               const ipc::IpcWriteOptions& options) override {
+    return Status::OK();
+  }
+
+  Status Begin(const std::shared_ptr<Schema>& schema) override {
+    return MetadataRecordBatchWriter::Begin(schema);
+  }
+
+  ipc::WriteStats stats() const override { return ipc::WriteStats(); }
+
+  Status WriteWithMetadata(const RecordBatch& batch,
+                           std::shared_ptr<Buffer> app_metadata) override {
+    return Status::OK();
+  }
+
+  Status Close() override { return Status::OK(); }
+
+  Status WriteRecordBatch(const RecordBatch& batch) override { return Status::OK(); }
+};
+
+FlightDescriptor getDescriptor(google::protobuf::Message& command) {
+  google::protobuf::Any any;
+  any.PackFrom(command);
+
+  const std::string& string = any.SerializeAsString();
+  return FlightDescriptor::Command(string);
+}
+
+TEST(TestFlightSqlClient, TestGetCatalogs) {
+  auto client_mock = std::make_shared<internal::FlightClientImpl>();
+  FlightSqlClient sql_client(client_mock);
```

Comment:
Hey @lidavidm ! Can you please clarify?

Reply:
We're defining FlightClientImpl twice and just using the fact that only one definition is visible at a time, trusting that the compiler/linker do the right thing. I'm not sure we can rely on this.

Reply:
I would expect that the object files for client.cc and client_test.cc both contain symbols for FlightClientImpl, and it just so happens that the linker picks the mock when compiling the test binary.

Reply:
That said, my understanding of this is shaky and maybe there is a good justification for why this works.

Reply:
The intent is that the library build and unit test build compile and link client.cc and client_test.cc respectively. These two files shouldn't be passed to the same linker run at once.

Reply:
Both libarrow_flight_sql.so and client_test.cc.o have symbols for FlightClientImpl for me. That said, they're marked as weak symbols, and presumably the one in the program takes precedence. (Confirmed with LD_DEBUG.)

This works and I'm not going to hold this up, however, it is rather magical and it's worth at least a comment explaining the intent of FlightClientImpl.

Reply:
We are defining FlightClientImpl on `client_test.cc` and `client_internal.cc`, made sure that we are not building or linking them at any time on CMakeLists.txt

Reply:
Er, sorry for the churn @rafael-telles but I think the latest commit is strictly worse. I think the situation is OK as-is so long as it's clearly documented (and I can't claim to be a linker expert or C++ standard expert so maybe this is all OK and I don't understand the reason why). I'm mostly just concerned that it's unclear what the intent is.

Reply:
Oops, didn't see your last comment!
I thought that as it was statically linking libarrow_flight_sql(containing client_internal.cc) when building client_test.cc it could be a problem.

Reply:
Perhaps, but even then, I would say that an implementation that requires very carefully linking objects together to be correct is not one we should go with.

Reply:
(Sorry - I want to try some alternative approaches but haven't found the time. At this rate it may be Friday before I can get into things…)

Reply:
Ok, sorry, can we avoid using `FlightClientImpl` like this? I don't think this will be maintainable/I think this will be brittle. We can do something like the following:

- Make `FlightSqlClient` an abstract base class, with an implementation `template <typename Client> class FlightSqlClientImpl : public FlightSqlClient { ... };` in `client_internal.h`. Then in `client.cc` we can define a factory method that instantiates and returns `FlightSqlClientImpl<FlightClient>` and in `client_test.cc` we can instantiate `FlightSqlClientImpl<MockedFlightClient>`. 

    This way the parameterization on the client impl is explicit and we are not relying on linker magic.
- Alternatively, instead of mocking, we can just implement in-process dummy servers to provide the expected reply. This is definitely more code but avoids having to parameterize the client.
- Alternatively, we can extract a `FlightClientInterface` or refactor `FlightClient` so it has virtual methods and is directly mockable.

Reply:
Hey @lidavidm , thank you for the suggestions!
I think we can do similar to the first suggestion:

Make `FlightSqlClient` as...
```c++
class ARROW_EXPORT FlightSqlClient {
 private:
  std::shared_ptr<FlightClient> client_;

 public:
  explicit FlightSqlClient(std::shared_ptr<FlightClient> client);

  ~FlightSqlClient();

  virtual Status GetFlightInfo(const FlightCallOptions& options,
                       const FlightDescriptor& descriptor,
                       std::unique_ptr<FlightInfo>* info) {
    return client_->GetFlightInfo(options, descriptor, info);
  }
  
  // Do the same with other FlightClient methods called
```

Then we can mock these virtual methods(only the ones we simply forward to FlightClient) from FlightSqlClient instead of FlightClient/FlightClientImpl on `client_test.cc`.

Reply:
That's a good idea! They could be made protected methods too.

Reply:
Done @lidavidm !